docopt/__init__.py

```diff
@@ -709,39 +709,101 @@ def isanumber(x):
     return parsed
 
 
-def parse_defaults(docstring: str) -> list[Option]:
-    defaults = []
-    for s in parse_section("options:", docstring):
-        options_literal, _, s = s.partition(":")
-        if " " in options_literal:
-            _, _, options_literal = options_literal.partition(" ")
-        assert options_literal.lower().strip() == "options"
-        split = re.split(r"\n[ \t]*(-\S+?)", "\n" + s)[1:]
-        split = [s1 + s2 for s1, s2 in zip(split[::2], split[1::2])]
-        for s in split:
-            if s.startswith("-"):
-                arg, _, description = s.partition("  ")
-                flag, _, var = arg.replace("=", " ").partition(" ")
-                option = Option.parse(s)
-                defaults.append(option)
-    return defaults
-
-
-def parse_section(name: str, source: str) -> list[str]:
-    pattern = re.compile(
-        "^([^\n]*" + name + "[^\n]*\n?(?:[ \t].*?(?:\n|$))*)",
-        re.IGNORECASE | re.MULTILINE,
+class DocSections(NamedTuple):
+    before_usage: str
+    usage_header: str
+    usage_body: str
+    after_usage: str
+
+
+def parse_docstring_sections(docstring: str) -> DocSections:
+    """Partition the docstring into the main sections.
+
+    The docstring is returned, split into a tuple of 4 pieces: text before the
+    usage section, the usage section header, the usage section body and text
+    following the usage section.
+    """
+    usage_pattern = r"""
+    # Any number of lines precede the usage section
+    \A(?P<before_usage>(?:.*\n)*?)
+    # The `usage:` section header.
+    ^(?P<usage_header>.*usage:)
+    (?P<usage_body>
+        # The first line of the body may follow the header without a line break:
+        (?:
+            # Some non-whitespace content
+            [ \t]*\S.*(?:\n|\Z)
+            # Or after a newline, followed by indentation
+            |(?:[ \t]*\n[ \t].*(?:\n|\Z))
+        )
+        # Any number of additional indented lines
+        (?:[ \t].*(?:\n|\Z))*
```

Comment:
This section of regex is confusing to me. Currently it's framed as "this is what constitues a usage body". Can we instead frame it as "everything is a usage body until we encounter the pattern that ends the usage body"?

Perhaps

```python
    usage_pattern = r"""
        # Any number of lines precede the usage section
        \A(?P<before_usage>
          (?:.*\n)*?
        )
        # The `usage:` section header.
        ^(?P<usage_header>
          .*\busage:
        )
        # Any number of lines
        (?P<usage_body>
            (?:.*(?:\n|\Z))*?
        )
        (?P<after_usage>
            # either nothing
            \Z
            # or...
            |
            (?:
                # The marker of the start of the next section
                # which is either an empty line or a non-indented line
                (?:\n|[^ \t])
                # then more lines
                (?:.*(?:\n|\Z))*
            )
        )
    """
```

Reply:
perhaps could use lookarounds here as well

Reply:
Perhaps this should even be done programatically instead of with regex? eg `for line in doc:...` It might be more clear that way.

Reply:
Yeah, on looking at it again I was a bit confused by how I'd implemented `usage_body` group. I think the only reason It had that "Some non-whitespace content ... Or after a newline, followed by indentation" was to prevent the usage_body group matching an empty usage section. i.e. the body had to either have something after immediately after the usage: header, or at least one indented line. That could be done more simply with a lookahead assertion.

I just tried your suggestion, it looks good intuitively, but unfortunately with the non-greedy match, `usage_body` always ends up being empty and the `after_usage` grabs everything, including the `usage_body`.

With some programatic validation to check the `usage_body` isn't empty, we could simplify the pattern it to this:
```python
    usage_pattern = r"""
    # Any number of lines (that don't include usage:) precede the usage section
    \A(?P<before_usage>(?:(?!.*\busage:).*\n)*)
    # The `usage:` section header.
    ^(?P<usage_header>.*\busage:)
    (?P<usage_body>
        # The first line of the body may follow the header without a line break:
        (?:.*(?:\n|\Z))
        # Any number of additional indented lines
        (?:[ \t].*(?:\n|\Z))*
    )
    # Everything else
    (?P<after_usage>(?:.|\n)*)\Z
    """
```

```python
if not match or not (match.group("usage_body").strip()):
    raise DocoptLanguageError(
        # ...
```
Could also move the responsibility for disallowing an empty usage body to `lint_docstring()` — allow `parse_docstring_sections()` to return an empty body, then in `lint_docstring()` raise an error specifically saying the body is empty.

I need to add a test for this actually, as the current implementation accepts a usage body containing only whitespace, and then triggers a `ValueError` when `formal_usage()` calls `.split()` on the whitespace string, producing only one value when it expects 2+.

Reply:
I've made this change I mention in the commits I've added, see what you think. Can easily change it again if you like.

Reply:
I prefer what you did!